runtime/tools/python/ttrt/common/memory.py

```diff
@@ -0,0 +1,82 @@
+# SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+import json
+import importlib.machinery
+import sys
+import signal
+import os
+import io
+import subprocess
+import time
+import socket
+from pkg_resources import get_distribution
+import shutil
+import atexit
+import re
+from functools import partial
+
+from ttrt.common.util import *
+
+
+class MemoryRuntimeConfig:
+    def __init__(self, device=None, artifact_dir="", tt_metal_home_path=""):
+        self.device = device
+        self.artifact_dir = artifact_dir
+        self.tt_metal_home_path = tt_metal_home_path
+
+
+def memory(memory_runtime_config, binary, program_context, op_context):
+    device = memory_runtime_config.device
+
+    print("-----------executing memory dump-----------")
+
+    try:
+        import ttrt.runtime
+        import ttrt.binary
+
+        device.dump_memory_report()
+
+        op_debug_str = ttrt.runtime.get_op_debug_str(op_context)
+
+        # find matching golden tensor based on loc in op debug string
+        match = re.search(r"loc\(([^)]+)\)", op_debug_str)
+
+        if not match:
+            print(f"debug_str={op_debug_str}")
+            print("No location found in debug string - skipping memory dump")
+            return
+
+        loc = match.group(1).replace('"', "")
+        print(f"found location={loc}")
+
+        match = re.search(r":(\d+:\d+)$", loc)
+
+        if not match:
+            print(f"loc_str={loc}")
+            print("No row:column found in loc string - skipping memory dump")
+            return
+
+        row_col = match.group(1)
+
+        memory_dump_dir_path = (
+            f"{memory_runtime_config.tt_metal_home_path}/generated/reports"
+        )
+        memory_dump_artifact_dir_path = (
+            f"{memory_runtime_config.artifact_dir}/{row_col}_memory_reports"
+        )
+        # os.mkdir(memory_dump_artifact_dir_path)
+        if os.path.exists(memory_dump_artifact_dir_path):
+            shutil.rmtree(memory_dump_artifact_dir_path)
+        shutil.copytree(memory_dump_dir_path, memory_dump_artifact_dir_path)
```

Comment:
Instead of moving dir trees around, maybe it's better to read the memory dump file and bring it into a common top level json file that's easier for us to use?

That way we can also just encode the loc / debug strings intact in the json instead of relying on row/col naming conventions.

Reply:
agreed. will fix